public/app/plugins/panel/graph/Legend/LegendSeriesItem.tsx

```diff
@@ -0,0 +1,211 @@
+import React from 'react';
+import { TimeSeries } from 'app/core/core';
+import withColorPicker from 'app/core/components/colorpicker/withColorPicker';
+
+export const LEGEND_STATS = ['min', 'max', 'avg', 'current', 'total'];
+
+export interface LegendLabelProps {
+  series: TimeSeries;
+  asTable?: boolean;
+  hidden?: boolean;
+  onLabelClick?: (series, event) => void;
+  onColorChange?: (series, color: string) => void;
+  onToggleAxis?: (series) => void;
+}
+
+export interface LegendValuesProps {
+  values?: boolean;
+  min?: boolean;
+  max?: boolean;
+  avg?: boolean;
+  current?: boolean;
+  total?: boolean;
+}
+
+type LegendItemProps = LegendLabelProps & LegendValuesProps;
+
+interface LegendItemState {
+  yaxis: number;
+}
+
+export class LegendItem extends React.PureComponent<LegendItemProps, LegendItemState> {
+  static defaultProps = {
+    asTable: false,
+    hidden: false,
+    onLabelClick: () => {},
+    onColorChange: () => {},
+    onToggleAxis: () => {},
+  };
+
+  constructor(props) {
+    super(props);
+    this.state = {
+      yaxis: this.props.series.yaxis,
+    };
+  }
+
+  onLabelClick = e => this.props.onLabelClick(this.props.series, e);
+
+  onToggleAxis = () => {
+    const yaxis = this.state.yaxis === 2 ? 1 : 2;
+    const info = { alias: this.props.series.alias, yaxis: yaxis };
+    this.setState({ yaxis: yaxis });
+    this.props.onToggleAxis(info);
+  };
+
+  onColorChange = color => {
+    this.props.onColorChange(this.props.series, color);
+  };
+
+  render() {
+    const { series, hidden, asTable } = this.props;
+    const seriesOptionClasses = getOptionSeriesCSSClasses(series, hidden);
```

Comment:
`getOptionSeriesCSSClasses` could be come either an util function for legend related stuff(if you plan to reuse it) or simply an instance method of this component. Util approach gives better unit-testability, but css classes are better to be tested in visual diffs, rather than unit tests. So I would suggest to implement this as an instance method